cmd/promtool/analyze.go

```diff
@@ -0,0 +1,334 @@
+		}
+	}
+	stats.avgPop = float64(totalPop) / float64(numSamples)
+	return stats, nil
+}
+
+func sortMatrix(matrix model.Matrix) {
+	sort.SliceStable(matrix, func(i, j int) bool {
+		return getLe(matrix[i]) < getLe(matrix[j])
+	})
+}
+
+func getLe(series *model.SampleStream) float64 {
+	lbs := model.LabelSet(series.Metric)
+	le, _ := strconv.ParseFloat(string(lbs["le"]), 64)
+	return le
+}
+
+func getBucketCountsAtTime(matrix model.Matrix, numBuckets, timeIdx int) ([]int, error) {
+	counts := make([]int, numBuckets)
+	counts[0] = int(matrix[0].Values[timeIdx].Value)
+	for i, bucket := range matrix[1:] {
+		curr := bucket.Values[timeIdx]
+		prev := matrix[i].Values[timeIdx]
+		// Assume the results are nicely aligned.
+		if curr.Timestamp != prev.Timestamp {
+			return counts, fmt.Errorf("matrix result is not time aligned")
+		}
+		counts[i+1] = int(curr.Value - prev.Value)
+	}
+	return counts, nil
+}
+
+type bucketBounds struct {
+	boundaries   int32
+	upper, lower float64
+}
+
+func makeBucketBounds(b *model.HistogramBucket) bucketBounds {
+	return bucketBounds{
+		boundaries: b.Boundaries,
+		upper:      float64(b.Upper),
+		lower:      float64(b.Lower),
+	}
+}
+
+func calcNativeBucketStatistics(series *model.SampleStream) (*statistics, error) {
+	stats := &statistics{
+		minPop: math.MaxInt,
+	}
+
+	overall := make(map[bucketBounds]struct{})
+	totalPop := 0
+	if len(series.Histograms) == 0 {
+		return nil, errNotNativeHistogram
+	}
+	if len(series.Histograms) == 1 {
+		return nil, errNotEnoughData
+	}
+	for _, histogram := range series.Histograms {
+		for _, bucket := range histogram.Histogram.Buckets {
+			bb := makeBucketBounds(bucket)
+			overall[bb] = struct{}{}
+		}
+		countPop := len(histogram.Histogram.Buckets)
+
+		totalPop += countPop
+		if stats.minPop > countPop {
+			stats.minPop = countPop
+		}
+		if stats.maxPop < countPop {
+			stats.maxPop = countPop
+		}
+	}
+	stats.avgPop = float64(totalPop) / float64(len(series.Histograms))
+	stats.total = len(overall)
+	return stats, nil
+}
+
+type distribution struct {
+	min, max, count int
+	avg             float64
+}
+
+func newDistribution() distribution {
+	return distribution{
+		min: math.MaxInt,
+	}
+}
+
+func (d *distribution) update(num int) {
+	if d.min > num {
+		d.min = num
+	}
+	if d.max < num {
+		d.max = num
+	}
+	d.count++
+	d.avg += float64(num)/float64(d.count) - d.avg/float64(d.count)
+}
+
+func (d distribution) String() string {
+	return fmt.Sprintf("min/avg/max: %d/%.3f/%d", d.min, d.avg, d.max)
+}
+
+type metaStatistics struct {
+	minPop, avgPop, maxPop, total distribution
+}
+
+func newMetaStatistics() *metaStatistics {
+	return &metaStatistics{
+		minPop: newDistribution(),
+		avgPop: newDistribution(),
+		maxPop: newDistribution(),
+		total:  newDistribution(),
+	}
+}
+
+func (ms metaStatistics) String() string {
+	return fmt.Sprintf("minPop - %v\navgPop - %v\nmaxPop - %v\ntotal - %v\ncount - %d", ms.minPop, ms.avgPop, ms.maxPop, ms.total, ms.minPop.count)
```

Comment:
I think this needs a bit more explanation to make sense to the user.

I would separate this part from the rest by a blank or so and then a header, similar to what I proposed above.
If we do the "unified mode", we should have two metaStatistics, one for native, one for classic histograms, and then print both if needed, but only one if only one is actually used.

Could look like this in total (including the proposal above):

```
Bucket stats for each series over time
--------------------------------------
First the min, avg, and max number of populated buckets, followed by
the total number of buckets (only for classic histograms).

- foo{bar="baz"} (classic): 3/12.882/14/21
- foo{bar="baz2"} (native): 3/122.787/154

Aggregated bucket stats
-----------------------
Each line shows min/avg/max over the series above.

Classic histograms (1 in total):
- min populated: 3/3.000/3
- avg populated: 12/12.000/12
- max populated: 14/14.000/14
- total: 21/21.000/21

Native histograms (1 in total):
- min populated: 3/3.000/3
- avg populated: 149/149.000/149
- max populated: 162/162.000/162
```

Reply:
Optional beautification challenge: You could aim for aligning the different fields (#12869 went down that path). But really just if you feel like it. Things could look like this then:

```Bucket stats for each series over time
--------------------------------------
First the min, avg, and max number of populated buckets, followed by
the total number of buckets (only for classic histograms).

- foo{bar="baz"}  (classic): 3/ 12.882/ 14/21
- foo{bar="baz2"} (native) : 3/122.787/154

Aggregated bucket stats
-----------------------
Each line shows min/avg/max over the series above.

Classic histograms (1 in total):
- min populated:  3/ 3.000/ 3
- avg populated: 12/12.000/12
- max populated: 14/14.000/14
- total:         21/21.000/21

Native histograms (1 in total):
- min populated:   3/  3.000/  3
- avg populated: 149/149.000/149
- max populated: 162/162.000/162
```